pkg/buildpack/managed_collection.go

```diff
@@ -55,16 +76,65 @@ func (f *ManagedCollection) AddModules(main BuildModule, deps ...BuildModule) {
 	}
 }
 
-// ShouldFlatten returns true if the given module is flattened.
-func (f *ManagedCollection) ShouldFlatten(module BuildModule) bool {
-	if f.flatten {
-		for _, modules := range f.flattenedModules {
-			for _, v := range modules {
-				if v == module {
-					return true
-				}
+func NewModuleManagerV2(modules FlattenModuleInfos) ManagedCollection {
+	flattenGroups := 0
+	if modules != nil {
+		flattenGroups = len(modules.FlattenModules())
+	}
+
+	return &managedCollectionV2{
+		flattenModuleInfos: modules,
+		managedCollection: managedCollection{
+			explodedModules:  []BuildModule{},
+			flattenedModules: make([][]BuildModule, flattenGroups),
+		},
+	}
+}
+
+// managedCollectionV2 can be used when flattenModuleInfos to be flattened are known beforehand. These flattenModuleInfos are provided during
+// initialization and the collection will take care of keeping them in the correct group once they are added.
+type managedCollectionV2 struct {
+	managedCollection
+	flattenModuleInfos FlattenModuleInfos
+}
+
+func (ff *managedCollectionV2) flattenGroups() []ModuleInfos {
+	return ff.flattenModuleInfos.FlattenModules()
+}
+
+func (ff *managedCollectionV2) AddModules(main BuildModule, deps ...BuildModule) {
+	var allModules []BuildModule
+	allModules = append(allModules, append([]BuildModule{main}, deps...)...)
+	for _, module := range allModules {
+		if ff.flattenModuleInfos != nil && len(ff.flattenGroups()) > 0 {
+			pos := ff.flattenGroup(module)
+			if pos >= 0 {
+				ff.flattenedModules[pos] = append(ff.flattenedModules[pos], module)
+			} else {
+				// this module must not be flattened
+				ff.explodedModules = append(ff.explodedModules, module)
 			}
+		} else {
+			// we don't want to flatten anything
+			ff.explodedModules = append(ff.explodedModules, module)
 		}
 	}
-	return false
+}
+
+// flattenGroup given a module it will try to determine to which row (group) this module must be added to in order to
+// be flattened. If it is not found, it means, the module must no me flattened at all
+func (ff *managedCollectionV2) flattenGroup(module BuildModule) int {
+	pos := -1
+	// flattenModuleInfos to be flattened are representing a two-dimension array. where each row represents a group of
+	// flattenModuleInfos that must be flattened together in the same layer.
+init:
+	for i, flattenGroup := range ff.flattenGroups() {
+		for _, buildModuleInfo := range flattenGroup.BuildModule() {
+			if buildModuleInfo.FullName() == module.Descriptor().Info().FullName() {
+				pos = i
+				break init
```

Comment:
```suggestion
				return i
```
Should do the same thing?